processor/internal/transformer/destination_transformer/logger.go

```diff
@@ -1,48 +1,61 @@
 package destination_transformer
 
 import (
+	"bytes"
+	"context"
 	"fmt"
+	"path"
 
 	"github.com/google/go-cmp/cmp"
 	"github.com/google/uuid"
-	"github.com/samber/lo"
 
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/logger"
 	obskit "github.com/rudderlabs/rudder-observability-kit/go/labels"
 
-	"github.com/rudderlabs/rudder-go-kit/stringify"
-
+	"github.com/rudderlabs/rudder-server/jsonrs"
 	"github.com/rudderlabs/rudder-server/processor/types"
-	"github.com/rudderlabs/rudder-server/utils/misc"
 )
 
 func (c *Client) CompareAndLog(
+	ctx context.Context,
 	embeddedResponse, legacyResponse types.Response,
 ) {
-	c.loggedEventsMu.Lock()
-	defer c.loggedEventsMu.Unlock()
+	if c.samplingFileManager == nil { // Cannot upload, we should just report the issue with no diff
+		c.log.Warnn("DestinationTransformer sanity check failed")
+		return
+	}
 
-	if c.loggedEvents >= int64(c.config.maxLoggedEvents.Load()) {
+	if c.loggedEvents.Load() >= int64(c.config.maxLoggedEvents.Load()) {
 		return
 	}
 
 	c.stats.comparisonTime.RecordDuration()()
 
 	differingResponse, sampleDiff := c.differingEvents(embeddedResponse, legacyResponse)
 	if len(differingResponse) == 0 && sampleDiff == "" {
-		c.log.Infof("Embedded and legacy responses are matches")
 		return
 	}
 
-	logEntries := lo.Map(differingResponse, func(item types.TransformerResponse, index int) string {
-		return stringify.Any(item)
-	})
-	if err := c.write(append([]string{sampleDiff}, logEntries...)); err != nil {
-		c.log.Warnn("Error logging events", obskit.Error(err))
+	objName := path.Join("embedded-dt-samples", config.GetKubeNamespace(), uuid.New().String())
+	differingResponseJSON, err := jsonrs.Marshal(differingResponse)
+	if err != nil {
+		c.log.Errorn("DestinationTransformer sanity check failed (cannot encode differingResponse)", obskit.Error(err))
 		return
 	}
 
-	c.log.Infof("Successfully logged events: %d", len(logEntries))
-	c.loggedEvents += int64(len(logEntries))
+	// upload sample diff and differing response to s3
+	file, err := c.samplingFileManager.UploadReader(ctx, objName, bytes.NewReader(append([]byte(sampleDiff), differingResponseJSON...)))
+	if err != nil {
+		c.log.Errorn("Error uploading DestinationTransformer sanity check diff file", obskit.Error(err))
+		return
+	}
+
+	c.log.Warnn("DestinationTransformer sanity check failed",
+		logger.NewStringField("location", file.Location),
+		logger.NewStringField("objectName", file.ObjectName),
+	)
+	c.loggedEvents.Add(int64(len(differingResponse)))
```

Comment:
This is probably coming from my own branch but on a second look I think this should be updated so that we bump `loggedEvents` as soon as we know how many events we got. Basically something like this:

```go
	differingResponse, sampleDiff := c.differingEvents(embeddedResponse, legacyResponse)
	if len(differingResponse) == 0 && sampleDiff == "" {
		return
	}

	c.loggedEvents.Add(int64(len(differingResponse)))
```

Reply:
The most precise I can think of is this: https://github.com/rudderlabs/rudder-server/pull/5816/files
Not overly complicated, also I don't think at this point we need a leaky logger either. The atomic approach seems safe and precise enough.